Create/Edit source secret

jhadvig opened the pull request:
Adding subform for creating source secret types:
- `kubernetes.io/basic-auth`
- `kubernetes.io/ssh-auth`

Screens:
![3](https://user-images.githubusercontent.com/1668218/41734702-04b66c50-7588-11e8-87a2-de0d7453ec1b.png)
![1](https://user-images.githubusercontent.com/1668218/41734706-07db5396-7588-11e8-85e8-c3d5704bc252.png)
![2](https://user-images.githubusercontent.com/1668218/41734704-069578ae-7588-11e8-8c7c-cea76197bbaf.png)

@openshift/team-ux-review PTAL

Going to do additional code refactoring

spadgett commented:
We might call it "Git Source Secret" to make it more obvious what it's used for.

benjaminapetersen commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -11,43 +11,62 @@ import { getActiveNamespace, formatNamespacedRouteForResource, UIActions } from
+}
+
+const determineDefaultSecretType = (typeAbstraction) => {
+  if (typeAbstraction === 'source'){
```
Just a style consistency nit:

```JavaScript 
return typeAbstraction === 'source' ? SecretType.basicAut : SecretType.opaque;
```

benjaminapetersen commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -72,17 +91,25 @@ class BaseEditSecret_ extends SafetyFirst<BaseEditSecretProps_, BaseEditSecretSt
 
-    const explanation = 'Webhook secrets allow you to authenticate a webhook trigger.';
-    const subform = <WebHookSecretSubform onChange={this.onDataChanged.bind(this)} stringData={this.state.stringData} />;
+    let explanation = null;
```
If you can `const` and separate the logic on these, I would:

```JavaScript 
const isWebhook = this.state.secretTypeAbstraction === 'webhook'; 
const explanation = isWebhook 
  ? 'Webhook secrets allow you to authenticate a webhook trigger.'
  : 'Source secrets allow you to authenticate against the SCM server.';
const subform = isWebhook
  ? <WebHookSecretSubform onChange={this.onDataChanged.bind(this)} stringData={this.state.stringData} />
  : <SourceSecretSubform onChange={this.onDataChanged.bind(this)} stringData={this.state.stringData} />;
```

spadgett replied:
There are going to be a bunch more subforms, so this won't scale. We'll eventually have a switch statement for several types.

benjaminapetersen replied:
Thats fine.  A `switch` or an `object` lookup works for me.

benjaminapetersen commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -139,25 +166,30 @@ const generateSecret = () => {
       <label className="control-label" htmlFor="webhook-secret-key">Webhook Secret Key</label>
       <div className="input-group">
-        <input className="form-control" id="webhook-secret-key" type="text" name="webhookSecretKey" onChange={this.changeWebHookSecretkey} value={this.state.WebHookSecretKey} required/>
+        <input className="form-control" id="webhook-secret-key" type="text" name="webhookSecretKey" onChange={this.changeWebHookSecretkey} value={this.state.stringData.WebHookSecretKey} required/>
```
Do we need to worry about ensuring `id` attribs are always unique, with something like `_.uniqueId('webhook-secret-key-')`?  We ran into this error a lot in the old console, where a component initially only appeared on the page once... then later would get reused & we would have `id` conflicts.  I'd suggest we just always ensure its unique.

benjaminapetersen commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -167,11 +199,93 @@ class WebHookSecretSubform extends React.Component<WebHookSecretSubformProps, We
+    }, () => this.props.onChange(this.state));
+  }
+  render () {
+    const basicAuthSubform = <React.Fragment>
```
Do these need to live inside `render`?  I'd prob rather favor smaller components & just pull them out as separate functions:

```JavaScript 
 const basicAuthSubform: React.SFC = ({props}) => <React.Fragment> ... </React.Fragment>
```
I hesitate to risk `render` bloating over time.

benjaminapetersen commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -167,11 +199,93 @@ class WebHookSecretSubform extends React.Component<WebHookSecretSubformProps, We
+          <input type="text" className="form-control" readOnly disabled/>
+          <span className="input-group-btn">
+            <span className="btn btn-default btn-file">
+              Browse&hellip;
```
Is `Browse&hellip;` correct?

jhadvig replied:
It should be -> https://github.com/openshift/origin-web-console/blob/efea73189b2e79f8923279b0935374c47d760942/app/views/directives/osc-file-input.html#L16

benjaminapetersen commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -167,11 +199,93 @@ class WebHookSecretSubform extends React.Component<WebHookSecretSubformProps, We
+          </select>
+        </div>
+      </div>
+      { this.state.authenticationType === 'kubernetes.io/basic-auth' ? basicAuthSubform : sshAuthSubform }
```
Per above comment, then:

```JavaScript
{ this.state.authenticationType === 'kubernetes.io/basic-auth' 
   ? <BasicAuthSubform {...theseProps} />
   : <SSHAuthSubform {...otherProps}> }
```

beanh66 commented:
@serenamarie125 Yes, agreed the dropdown is not aligned correctly, but I think that is a known issue and @jhadvig was addressing in another PR. Correct me if I'm wrong @jhadvig!

jhadvig commented:
Yeah its a known issue, I've described it [here](https://github.com/openshift/console/pull/94#issuecomment-397229150). Will open an issue for that, so we can track it. 
@serenamarie125 @beanh66 any other suggestions ? Thanks ! :)

robszumski commented:
Jenkins push

jhadvig commented:
@spadgett @benjaminapetersen I've refactored and updated the PR and the `create-secret.tsx` file.

Refactoring includes:
- Properly naming class and function componnets
- Adding a Higher Order Component - `withSecretForm` - which is responsible or constructing different secret forms (source, webhook) and returns `SecretFormComponent` that represents particular form 

The overall component hierarchy looks like:
```
SecretFormComponent --  WebHookSecretForm
                    \
                     \  SourceSecretForm  --  BasicAuthSubform
                                          \
                                           \  SSHAuthSubform  -- FileInput
```
Also added:
- `FileInput` component that is responsible for uploading files

Had created support for drag'n'drop, but since this PR already has a lot of changes, so I will submit the PR after this is merged.  

PTAL

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -2,52 +2,77 @@
+  return s4() + s4() + s4() + s4();
+};
+
+// withSecretForm returns SubformComponent which is a Higher Order Component for all the types of secret forms.
```
It returns the SecretFormComponent, right? Not SubformComponent?

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -59,116 +84,258 @@ class BaseEditSecret_ extends SafetyFirst<BaseEditSecretProps_, BaseEditSecretSt
         <title>{title}</title>
       </Helmet>
-      <form className="co-m-pane__body-group" onSubmit={this.save}>
+      <form className="co-m-pane__body-group create-secret-form" onSubmit={this.save}>
```
Prefer to prefix class names to avoid collisions with dependencies and for consistency with existing styles: `co-create-secret-form`

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -2,52 +2,77 @@
-    this.setState({stringData: {...secretsData}});
+    this.setState({
+      stringData: {...secretsData.stringData},
+      type: secretsData.authenticationType,
```
It sort of breaks the abstraction if the SecretFormComponent knows about authenticationType. Can this be changed to just "type"?

jhadvig replied:
makes sense

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -59,116 +84,258 @@ class BaseEditSecret_ extends SafetyFirst<BaseEditSecretProps_, BaseEditSecretSt
+  constructor (props) {
+    super(props);
+    this.state = {
+      authenticationType: this.props.secretType,
```
Yeah, I suggest just making this `type`, then it can be handled generically as type by `SecretFormComponent`

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -59,116 +84,258 @@ class BaseEditSecret_ extends SafetyFirst<BaseEditSecretProps_, BaseEditSecretSt
+    return <React.Fragment>
+      {this.props.isCreate
+        ? <div className="form-group">
+          <label className="control-label">Authentication Type</label>
```
Can you add htmlFor?

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -59,116 +84,258 @@ class BaseEditSecret_ extends SafetyFirst<BaseEditSecretProps_, BaseEditSecretSt
+          <div className="modal-body__field">
+            <select onChange={this.changeAuthenticationType} value={this.state.authenticationType} className="form-control">
+              <option key="kubernetes.io/basic-auth" value="kubernetes.io/basic-auth">Basic Authentication</option>
+              <option key="kubernetes.io/ssh-auth" value="kubernetes.io/ssh-auth">SSH Key</option>
```
I don't think `key` attributes are needed?

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -59,116 +84,258 @@ class BaseEditSecret_ extends SafetyFirst<BaseEditSecretProps_, BaseEditSecretSt
+      {this.props.isCreate
+        ? <div className="form-group">
+          <label className="control-label">Authentication Type</label>
+          <div className="modal-body__field">
```
Avoid using a modal-body style outside a modal. This doesn't align with BEM

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -59,116 +84,258 @@ class BaseEditSecret_ extends SafetyFirst<BaseEditSecretProps_, BaseEditSecretSt
+      }
+      { this.state.authenticationType === 'kubernetes.io/basic-auth'
+        ? <BasicAuthSubform onChange={this.onDataChanged.bind(this)} stringData={this.state.stringData}/>
+        : <SSHAuthSubform onChange={this.onDataChanged.bind(this)} stringData={this.state.stringData}/>
```
Prefer to `bind` in the constructor

nit: space before self-closing `/`

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -59,116 +84,258 @@ class BaseEditSecret_ extends SafetyFirst<BaseEditSecretProps_, BaseEditSecretSt
+    return <React.Fragment>
+      <div className="form-group">
+        <label className="control-label" htmlFor="username">Username</label>
+        <div className="modal-body__field">
```
Same comment on modal-body styles

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -59,116 +84,258 @@ class BaseEditSecret_ extends SafetyFirst<BaseEditSecretProps_, BaseEditSecretSt
+    return <div className="form-group">
+      <label className="control-label" htmlFor="ssh-privatekey">SSH Private Key</label>
+      <div className="modal-body__field">
+        <FileInput onChange={this.onFileChange.bind(this)}/>
```
Prefer to bind in constructor.

nit: space before `/` in self-closing tags since that's the style used throughout the existing code

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -59,116 +84,258 @@ class BaseEditSecret_ extends SafetyFirst<BaseEditSecretProps_, BaseEditSecretSt
+          name="privateKey"
+          onChange={this.changeData}
+          value={this.state['ssh-privatekey']}
+          required>
```
nit: self-closing tag?

spadgett replied:
aria-describedby would be good here

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -182,18 +349,48 @@ export type BaseEditSecretProps_ = {
+export type WebHookSecretFormProps = {
   onChange: Function;
-  stringData: {[WebHookSecretKey: string]: string};
+  stringData: {[WebHookSecretKey: string]: string},
```
Missed this only previous review, but this should be

```ts
stringData: {
  WebHookSecretKey: string,
}
```

spadgett commented on frontend/public/components/utils/file-input.tsx:
```diff
@@ -0,0 +1,50 @@
+import * as React from 'react';
+
+export class FileInput extends React.Component<fileInputProps, fileInputState> {
```
👍 nice to see this added with so little code

jhadvig replied:
yeah, even the drag&Drop is pretty small :+1:

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -2,52 +2,77 @@
+};
+
+// withSecretForm returns SubformComponent which is a Higher Order Component for all the types of secret forms.
+const withSecretForm = (SubformComponent) => class SecretFormComponent extends React.Component<BaseEditSecretProps_, BaseEditSecretState_> {
```
I think you can take "Component" out of the name. I'd call this `class SecretForm`

jhadvig replied:
Make even more sense ! :)

spadgett commented:
@jhadvig This PR is big enough that it would be good to address feedback in separate commits, then we can squash before merge. Then it's easier to review just what's been updated.

jhadvig commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -2,52 +2,77 @@
-    const secret = _.defaultsDeep({}, props.fixed, existingObj, {
+    const inputObj = _.get(props.obj, 'data');
+    const existingSecret = _.pick(inputObj, ['metadata', 'type']);
+    const defaultSecretType = determineDefaultSecretType(this.props.secretTypeAbstraction);
```
So this is for when user creates secret in a form where he has to select from a secret type, [eg.](https://github.com/jhadvig/console/blob/0a265d663a0764d51b749b4dcab294419670fda6/frontend/public/components/secrets/create-secret.tsx#L202-L204) In that case a default value has to be set for the *state.secret*. 
The image form will probably pick from image secret types - `kubernetes.io/dockercfg` & `kubernetes.io/dockerconfigjson`

jhadvig commented on frontend/public/components/_secret.scss:
```diff
@@ -0,0 +1,29 @@
+.btn-file {
```
So in case we want the input field with `Browse` btn for picking file, we need additional styling since the Bootstrap does not offer this styling by default.
Check origin-web-console [styling](https://github.com/openshift/origin-web-console/blob/master/app/styles/_buttons.less#L3-L19) and [usage](https://github.com/openshift/origin-web-console/blob/master/app/views/directives/osc-file-input.html#L5-L20). Also was browsing similar use-cases and more less this approach is pretty common.

jhadvig commented:
@spadgett comments addressed in a separate [commit](https://github.com/openshift/console/pull/150/commits/fbcd8cdab9f1c5e83c22d0c9616d339ca89ddad6) and also commented back on two of the comments:
- https://github.com/openshift/console/pull/150#discussion_r198071270
- https://github.com/openshift/console/pull/150#discussion_r197962238

PTAL

spadgett commented on frontend/public/components/_secret.scss:
```diff
@@ -0,0 +1,31 @@
+.co-create-secret-form {
+  .btn-file {
```
The file picker is generic and will eventually be used outside the secret form. We shouldn't scope the styles to the secret form.

Please use `co-` prefix for new styles.

spadgett commented on frontend/public/components/_secret.scss:
```diff
@@ -0,0 +1,31 @@
+.co-create-secret-form {
```
BEM discourages nesting, and we should give the classes appropriate names for BEM.

http://getbem.com/

spadgett replied:
Style properties should be alphabetized.

jhadvig replied:
Saw nesting in our code-base so thought it's ok, but will try to avoid it ! Thanks :)

spadgett commented on frontend/public/components/_secret.scss:
```diff
@@ -0,0 +1,31 @@
+  }
+
+  .help-block {
+    margin-bottom: 10px;
```
Why do we need to override the default margin for help-block?

jhadvig commented:
@spadgett comments addressed. Thanks and PTAL :)

spadgett commented on frontend/public/components/utils/_file-input.scss:
```diff
@@ -0,0 +1,19 @@
+    outline: none;
+    background: white;
+    cursor: inherit;
+    display: block;
```
nit: alphabetize

jhadvig commented:
@spadgett comment addressed and squashed. Lets merge this baby :+1:

spadgett commented:
/refresh

stevekuznetsov commented:
/retest
